### DISCUSSION TITLE
MCUBOOT_DIRECT_XIP mode

### DISCUSSION THREAD

**Discussion by Sathishkumar-rk** [zephyrproject-rtos/zephyr#48259] [Q&A] [answered]
Hi @DerekSnell @danieldegrasse,

Can you please give your thoughts on below request.

I want to use **MCUBOOT_DIRECT_XIP** upgrade mode to run latest image from their slots.

Trials for **MCUBOOT_DIRECT_XIP**:

1.  Slot 0 with latest image
- [ ] Flashed slot-0 with image version 3.3.3+3
- [ ] Flashed slot-1 with image version 2.2.2+2
  Output:
   **Slot 0 image is getting executed**

2. Slot 1 with latest image
 - [ ]  Flashed slot-0 with image version 2.2.2+2
 - [ ] Flashed slot-1 with image version 3.3.3+3 
 Output:
    **Slot 0 image getting executed** but expectation is Slot 1 image will get execute.

I'm expecting in second test case from slot-1 image should get executed as per documentation details, whichever slot has higher version should get executed. below is my configuration details of MCUBoot and Zephyr application image.

- MCUBoot configuration
![image](https://user-images.githubusercontent.com/105291276/180771660-4c580bcc-c58f-40aa-9dda-ad98517519a1.png)

- Image preparation 
![image](https://user-images.githubusercontent.com/105291276/180771881-81ad0857-3998-4289-bc4d-3246b5939993.png)
![image](https://user-images.githubusercontent.com/105291276/180772829-a4d7f0f4-109c-4ea1-a0eb-84871d6e58b0.png)

**Comment by DerekSnell** [comment]
Hi @Sathishkumar-rk ,

The image in Slot1 needs to be confirmed.  See steps [using MCUmgr](https://github.com/zephyrproject-rtos/zephyr/discussions/45252), and [MCUboot documentation](https://docs.mcuboot.com/design.html).

**Reply by Sathishkumar-rk** [reply]
Hi @DerekSnell

I have tried steps provided in zephyr_MCUBoot_Lab_v1.0.pdf file to upgrade the image from mcumgr. 
![image](https://user-images.githubusercontent.com/105291276/189047102-73bb2874-9b21-41ec-bd31-bdcf95de283f.png)

![image](https://user-images.githubusercontent.com/105291276/189047257-f5e9273b-e4c3-405d-bcf1-93adaedea995.png)

while running image list command from mcumgr receiving NMP timeout error response.

Is there any other steps Am I missing in building process like should I change any MCUBoot settings to adopt mcumgr functions?

Thanks,
Sathishkumar K

**Comment by DerekSnell** [comment, marked as answer]
Hi @Sathishkumar-rk ,
I have not run into NMP timeout errors before.  But after reviewing [this other issue](https://github.com/zephyrproject-rtos/zephyr/discussions/46003), and reading this [note about NMP timeouts](https://docs.zephyrproject.org/latest/samples/subsys/mgmt/mcumgr/smp_svr/README.html#upload-the-signed-image) in the smp_svr doc, it seems the mcumgr host tool will report this error when it can no longer communicate with smp_svr over the serial connection.  

Since this happens to you after giving the image list command, I would use a debugger with the smp_svr app, and step through the code after it receives this command, to see why the app stops responding.

Let us know what you find.

**Reply by Sathishkumar-rk** [reply]
Hi @DerekSnell 

Thanks for the update.

Let me check the commands, which is given on community thread.

Will update the status here after debugging.

Thank you.

**Reply by Sathishkumar-rk** [reply]
Hi @DerekSnell 

I can able to run the mcumgr with RT board.

From the issue forum which you have shared in last reply, found out in dts file uart port missing for **mcumgr** 

<img width="301" alt="image" src="https://user-images.githubusercontent.com/105291276/191050449-3c7ce43c-0239-48a6-b611-c7377f1b0935.png">

with this uart port inclusion I cab able to use **mcumgr**

Thanks for the support. I will confirm with these changes whether DIRECT_XIP mode is working as expected or not & will update the status.

Thanks,
Sathishkumar K

**Comment by Sathishkumar-rk** [comment]
Hi @DerekSnell,

I have tested secondary slot image with confirm command but still my expectation is not met with confirm image.

![image](https://user-images.githubusercontent.com/105291276/206664202-8869ab5c-b08d-4d76-9513-297cc35d33c5.png)

In above log information, chain-load offset address pointing to secondary slot location it's as expected and from their application reset vector is called as application entry point.

Actually from debugging I understood that, when we do west build it's preparing application binary for primary slot. which is confirmed by zephyr memory map file. Because of this even-though secondary slot image calling application reset vector memory it's pointing out to primary location that's why it's always running primary location image.

Is that any way we can build application for secondary slot location, so that upon reset vector call it will point to secondary image entry point.

Earlier I was expected like if binary image flashed over secondary slot MCUboot direct XIP mode will take care of address related stuff but it seems it not getting handled.

To confirm the above statement i was tried to change the flash base address of application through menuconfig setting and build application binary. Flashed in secondary slot and confirmed.

**Log Information:
binary build with default configuration:**
*** Booting Zephyr OS build b6270b6acb3b  ***
I: Starting Direct-XIP bootloader
I: Current datetime: 2022-11-14 19:00:00
I: Primary   slot: version=1.1.1+1 Header Info=0x96f3b83d, 0x0, 0x400, 0x0, 0xf21c, 0x0, 0x0
I: Secondary slot: version=2.2.2+2 Header Info=0x96f3b83d, 0x0, 0x400, 0x0, 0xf21c, 0x0, 0x0
I: Image 0 loaded from the secondary slot
I: Bootloader chainload address offset: **0x320000**
I: Jumping to the image slot
*** Booting Zephyr OS build b6270b6acb3b  ***
[00:00:00.004,089] <inf> smp_sample: build time: Nov 23 2022 23:41:44
[00:00:00.004,119] <inf> smp_sample: version 1.1.1+1

**Binary with modified configuration for secondary slot**
![image](https://user-images.githubusercontent.com/105291276/206668040-be3fee80-6e7d-486c-b5e4-f58c9bffd556.png)

*** Booting Zephyr OS build b6270b6acb3b  ***
I: Starting Direct-XIP bootloader
I: Current datetime: 2022-11-14 19:00:00


I: Primary   slot: version=1.1.1+1 Header Info=0x96f3b83d, 0x0, 0x400, 0x0, 0xf21c, 0x0, 0x0
I: Secondary slot: version=2.2.2+2 Header Info=0x96f3b83d, 0x0, 0x400, 0x0, 0xf21c, 0x0, 0x0
I: Image 0 loaded from the secondary slot
I: Bootloader chainload address offset: **0x320000**
I: Jumping to the image slot
*** Booting Zephyr OS build b6270b6acb3b  ***
[00:00:00.004,089] <inf> smp_sample: build time: Dec  9 2022 12:32:43
[00:00:00.004,119] <inf> smp_sample: version 2.2.2+2

Thanks,
Sathishkumar K
[zephyr_build_mapfile.zip](https://github.com/zephyrproject-rtos/zephyr/files/10193442/zephyr_build_mapfile.zip)
[mimxrt1160_evk_cm7.zip](https://github.com/zephyrproject-rtos/zephyr/files/10193453/mimxrt1160_evk_cm7.zip)

**Comment by DerekSnell** [comment]
Hi @Sathishkumar-rk ,

I am no expert on MCUboot.  My experience with MCUboot has always executed the applications from Slot0, and using MCUboot to swap the image in Slot1 to Slot0 before executing.

But if you want the linker to place the application code in the Slot1 partition for executing there, you just need to change the devicetree for the board.  The [line below is from the board's devicetree file](https://github.com/zephyrproject-rtos/zephyr/blob/1aebcec02ff9b901cab8056bc6168266033efe0a/boards/arm/mimxrt1160_evk/mimxrt1160_evk_cm7.dts#L25), and normally places code in the Slot0 partition.  The change below will move the executable image to Slot1.

```
chosen {
	zephyr,code-partition = &slot1_partition;
};
```

Best regards

**Reply by Sathishkumar-rk** [reply]
Hi @DerekSnell,

It makes sense to my understanding, Thanks for the information.
